File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(strings, predicate, /):
  for str in strings:
    if predicate(str): 
      return str.lower()
  return None

# Write your powers generator here
def powers_generator(base, limit):
    value = 1
    exponent = 0
    
    while value <= limit:
        yield value
        exponent += 1
        value = base ** exponent

# Write your say function here
def say(word=None):
    if word is None:
        return ''

    def inner_say(next_word=None, words=[word]):
        if next_word is None:
            return ' '.join(words)
        else:
            words.append(next_word)
            return lambda w=None: inner_say(w, words)
    return inner_say

# Write your line count function here
def meaningful_line_count(file, /):
    try:
        count = 0
        with open(file, 'r') as file:
            for line in file:
                stripped_line = line.strip()
                if stripped_line and not stripped_line.startswith('#'):
                    count += 1
            return count
    except FileNotFoundError as e:
        raise 
    
    
class Quaternion:
    def __init__(self, a: float, b: float, c: float, d: float):
        self.a = a
        self.b = b
        self.c = c
        self.d = d

    # String representation
    def __str__(self):
        parts = []
        if self.a != 0 or (self.a == 0 and self.b == 0 and self.c == 0 and self.d == 0):
            parts.append(f"{self.a}")
        if self.b != 0:
            parts.append(f"{'+' if self.b > 0 else ''}{self.b}i")
        if self.c != 0:
            parts.append(f"{'+' if self.c > 0 else ''}{self.c}j")
        if self.d != 0:
            parts.append(f"{'+' if self.d > 0 else ''}{self.d}k")
        return ''.join(parts).replace("+-", "-") if parts else '0'

    # Add two quaternions
    def __add__(self, other: 'Quaternion') -> 'Quaternion':
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)

    # Multiply two quaternions
    def __mul__(self, other: 'Quaternion') -> 'Quaternion':
        return Quaternion(
            self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
            self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
            self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
            self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        )

    # Conjugate 
    @property
    def conjugate(self) -> 'Quaternion':
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    # Coefficients 
    @property
    def coefficients(self) -> tuple:
        return (self.a, self.b, self.c, self.d)

    



# Write your Quaternion class here

File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(strings, predicate)
  for index, str in ipairs(strings) do
    if predicate(str) then
      return string.lower(str)
    end
  end
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
    return coroutine.create(function()
        local value = 1
        local exponent = 0
        while value <= limit do
            coroutine.yield(value)
            exponent = exponent + 1
            value = base ^ exponent
        end
    end)
end
-- Write your say function here
function say(word)
  local words = {}
  local function addWord(str)
  if str then
    table.insert(words, str)
    return addWord
  else
    return table.concat(words, " ")
  end
  end

  if word then
    return addWord(word)
  else
    return ""
  end
end

-- Write your line count function here
function meaningful_line_count(file)
  local count = 0
  local f, err = io.open(file, "r")
  if not f then
    error("File not found: " .. err)
  end
  for line in f:lines() do
    local trimmed_line = line:match("^%s*(.-)%s*$") -- lloks funky but removes whitespace
    if trimmed_line ~= "" and trimmed_line:sub(1,1) ~= "#" then -- checks first char of first line
      count = count + 1
    end
  end
  f:close()
  return count
end

-- Write your Quaternion table here
Quaternion = {}
Quaternion.__index = Quaternion

-- Constructor
function Quaternion.new(a, b, c, d)
    local self = setmetatable({}, Quaternion)
    self.a = a or 0  -- real part
    self.b = b or 0  -- i component
    self.c = c or 0  -- j component
    self.d = d or 0  -- k component
    return self
end

-- String representation of quaternion
function Quaternion:__tostring()
    local parts = {}
    if self.a ~= 0 or (self.a == 0 and self.b == 0 and self.c == 0 and self.d == 0) then
        table.insert(parts, tostring(self.a))
    end
    if self.b ~= 0 then
        table.insert(parts, (self.b == 1 and "" or self.b == -1 and "-" or self.b) .. "i")
    end
    if self.c ~= 0 then
        table.insert(parts, (self.c == 1 and "" or self.c == -1 and "-" or self.c) .. "j")
    end
    if self.d ~= 0 then
        table.insert(parts, (self.d == 1 and "" or self.d == -1 and "-" or self.d) .. "k")
    end
    return table.concat(parts, "+"):gsub("%+%-", "-")
end

-- Conjugate of a quaternion
function Quaternion:conjugate()
    return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

-- Addition of two quaternions
function Quaternion.__add(q1, q2)
    return Quaternion.new(q1.a + q2.a, q1.b + q2.b, q1.c + q2.c, q1.d + q2.d)
end

-- Multiplication of two quaternions 
function Quaternion.__mul(q1, q2)
    return Quaternion.new(
        q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,
        q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,
        q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,
        q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
    )
end

-- Coefficients list
function Quaternion:coefficients()
    return {self.a, self.b, self.c, self.d}
end

-- Equality check 
function Quaternion.__eq(q1, q2)
    return q1.a == q2.a and q1.b == q2.b and q1.c == q2.c and q1.d == q2.d
end

return Quaternion

